File: src/check/check.py
import os
import requests
import time
import threading
from threading import Thread
from helpers import take_new_screenshot, get_old_screenshot, update_old_screenshot, notify, get_difference, update_last_run, cleanup

# Difference tolerence
threshold = 0
now = int(time.time())


production = 'KUBERNETES_SERVICE_HOST' in os.environ

if production:
    configure_host = os.environ['CONFIGURE_SERVICE_HOST']
    configure_port = os.environ['CONFIGURE_SERVICE_PORT']
    configuration_address = f'http://{configure_host}:{configure_port}'
else:
    configure_address = 'http://0.0.0.0:8004'

def run_watcher(watcher_id, data):
    user_id = data['user_id']
    last_run = data['last_run']
    frequency = data['frequency']
    url = data['url']

    seconds_since_last_run = now - last_run
    if seconds_since_last_run > frequency:
        start = time.perf_counter()

        print(f'{watcher_id}: Updating')

        # Concurrently grab new and old screenshots
        server_file_paths = {
            'old' : None,
            'new' : None
        }
        # Take new screenshot
        t1 = Thread(target = take_new_screenshot, args = (watcher_id, url, server_file_paths, ))
        # Get old screenshot
        t2 = Thread(target = get_old_screenshot, args = (watcher_id, server_file_paths, ))
        t1.start()
        t2.start()
        t1.join()
        t2.join()

        # Get difference and notify if appropriate
        difference = get_difference(server_file_paths)

        if difference > threshold:
            print(f'{watcher_id}: Difference detected')

            # Notify
            t = Thread(target=notify, args=(user_id, url, ))
            t.start()
            
            # Update COS image
            t = Thread(target=update_old_screenshot, args=(server_file_paths['new'],))
            t.start()


        # Update last_run time  
        t = Thread(target=update_last_run, args=(watcher_id, now,))
        t.start()
        # Cleanup
        t = Thread(target=cleanup, args=(server_file_paths,))
        t.start()

        print(f'{watcher_id}: Finished in {time.perf_counter() - start} seconds')
        
    else:
        print(f'{watcher_id}: {seconds_since_last_run} < {frequency}')



if __name__ == "__main__":

    print('-----------------------------------------------------------')
    print(f'Epoch Time: {now}, Main thread: {threading.get_ident()}')
    print('-----------------------------------------------------------')

    watchers = requests.get(f'{configure_address}/watchers').json()

    threads = []
    for watcher_id, data in watchers.items():
        t = Thread(target=run_watcher, args=(watcher_id, data, ))
        threads.append((watcher_id, t))
        t.start()


    for (_,thread) in threads:
        thread.join()

    print("Main thread finished")

File: src/check/check.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

var now = time.Now().Unix()
var threshold = 0.02
var cloudObjectStorageAddress = "http://0.0.0.0:8001"
var compareAddress = "http://0.0.0.0:8002"
var screenshotAddress = "http://0.0.0.0:8003"
var configureAddress = "http://0.0.0.0:8004"
var notifyAddress = "http://0.0.0.0:8006"

// func handleWatcher(wid string, data map[string]interface{}, wg *sync.WaitGroup) {
// 	defer wg.Done()
func handleWatcher(wid string, data map[string]interface{}) {

	lastRun := int64(data["last_run"].(float64))
	frequency := int64(data["frequency"].(float64))

	if now > lastRun+frequency {
		// Get old SS, get new SS, can get both files at same time in seperate threads
		var wg2 sync.WaitGroup
		var url string
		// Retrieve from COS
		wg2.Add(1)
		COSFilePath := "files/COS" + wid + ".png"
		url = cloudObjectStorageAddress + "/files/" + wid + ".png"
		fmt.Println(wid, "fetching old", url)
		go getRequestToFile(url, COSFilePath, nil, &wg2)
		fmt.Println(wid, "done old", url)
		defer os.Remove(COSFilePath)
		// Take new SS
		wg2.Add(1)
		SSFilePath := "files/" + wid + ".png"
		url = screenshotAddress + "/screenshot"
		watcherURL := data["url"].(string)
		payload := strings.NewReader("{\n   \"url\":\"" + watcherURL + "\"\n}")
		fmt.Println(wid, "fetching new", watcherURL)
		go getRequestToFile(url, SSFilePath, payload, &wg2)
		fmt.Println(wid, "done new", watcherURL)
		defer os.Remove(SSFilePath)
		wg2.Wait()
		////////////////////////////////////////

		// Get difference, notify user if over threshold
		url = compareAddress + "/difference"
		d := getDifference(url, COSFilePath, SSFilePath)
		fmt.Println(wid, " difference: ", d)
		if d >= threshold {
			// Get bounding box photo
			DFilePath := "files/D" + wid + ".png"
			url = compareAddress + "/difference_image"
			getDifferenceImage(url, COSFilePath, SSFilePath, DFilePath)
			defer os.Remove(DFilePath)

			// Mail with boundind box photo
			url = configureAddress + "/users"
			email := data["email"].(string)
			fmt.Println(wid, "difference found, mailing", email)
			url = notifyAddress + "/notify"
			go notifyUser(url, email, watcherURL, DFilePath)

			// Upload new SS
			go uploadCOS(cloudObjectStorageAddress+"/files", SSFilePath)
		}

		url = configureAddress + "/watchers/" + wid + "?last_run=" + strconv.FormatInt(now, 10)
		updateWatcher(url)

	} else {
		fmt.Println(wid, "< frequency")
	}

	fmt.Println(wid, "thread completed")
}

func main() {

	// Check if on K8 cluster and we should be getting address from env variables for services
	_, production := os.LookupEnv("KUBERNETES_SERVICE_HOST")
	if production {
		cloudObjectStorageAddress = getProdServiceAddress("CLOUD_OBJECT_STORAGE")
		compareAddress = getProdServiceAddress("COMPARE")
		screenshotAddress = getProdServiceAddress("SCREENSHOT")
		configureAddress = getProdServiceAddress("CONFIGURE")
		notifyAddress = getProdServiceAddress("NOTIFY")
	}

	fmt.Println("--------------------------------------------------------------------------------------------------------------------")
	fmt.Println("COS service address:", cloudObjectStorageAddress)
	fmt.Println("Compare service address:", compareAddress)
	fmt.Println("Screenshot service address:", screenshotAddress)
	fmt.Println("Configure service address:", configureAddress)
	fmt.Println("Notify service address:", notifyAddress)
	fmt.Println("Start time:", now)
	fmt.Println("--------------------------------------------------------------------------------------------------------------------")

	watchers := getRequestToInterface(configureAddress + "/watchers").(map[string]interface{})

	// var wg sync.WaitGroup
	for wid, data := range watchers {
		// wg.Add(1)
		data := data.(map[string]interface{})
		handleWatcher(wid, data)
		// go handleWatcher(wid, data, &wg)
	}
	// wg.Wait()
	fmt.Println("Main thread completed")
	fmt.Println("--------------------------------------------------------------------------------------------------------------------")

}
